corehq/apps/sms/models.py

```diff
@@ -656,3 +690,487 @@ def opt_out_sms(cls, phone_number):
+
+        obj = MessagingSubEvent.objects.create(
+            parent=self,
+            date=datetime.utcnow(),
+            recipient_type=recipient_type,
+            recipient_id=recipient.get_id if recipient_type else None,
+            content_type=content_type,
+            form_unique_id=form_unique_id,
+            form_name=form_name,
+            case_id=reminder.case_id,
+            status=MessagingEvent.STATUS_IN_PROGRESS,
+        )
+        return obj
+
+    def create_ivr_subevent(self, recipient, form_unique_id, case_id=None):
+        recipient_type = MessagingEvent.get_recipient_type(recipient)
+        obj = MessagingSubEvent.objects.create(
+            parent=self,
+            date=datetime.utcnow(),
+            recipient_type=recipient_type,
+            recipient_id=recipient.get_id if recipient_type else None,
+            content_type=MessagingEvent.CONTENT_IVR_SURVEY,
+            form_unique_id=form_unique_id,
+            form_name=MessagingEvent.get_form_name_or_none(form_unique_id),
+            case_id=case_id,
+            status=MessagingEvent.STATUS_IN_PROGRESS,
+        )
+        return obj
+
+    @classmethod
+    def create_event_for_adhoc_sms(cls, domain, recipient=None,
+            content_type=CONTENT_ADHOC_SMS, source=SOURCE_OTHER):
+        if recipient:
+            recipient_type = cls.get_recipient_type(recipient)
+            recipient_id = recipient.get_id
+        else:
+            recipient_type = cls.RECIPIENT_UNKNOWN
+            recipient_id = None
+
+        obj = cls.objects.create(
+            domain=domain,
+            date=datetime.utcnow(),
+            source=source,
+            content_type=content_type,
+            status=cls.STATUS_IN_PROGRESS,
+            recipient_type=recipient_type,
+            recipient_id=recipient_id,
+        )
+        return obj
+
+    def create_structured_sms_subevent(self, case_id):
+        obj = MessagingSubEvent.objects.create(
+            parent=self,
+            date=datetime.utcnow(),
+            recipient_type=self.recipient_type,
+            recipient_id=self.recipient_id,
+            content_type=MessagingEvent.CONTENT_SMS_SURVEY,
+            form_unique_id=self.form_unique_id,
+            form_name=self.form_name,
+            case_id=case_id,
+            status=MessagingEvent.STATUS_IN_PROGRESS,
+        )
+        return obj
+
+    def create_subevent_for_single_sms(self, recipient_doc_type=None, recipient_id=None, case=None):
+        obj = MessagingSubEvent.objects.create(
+            parent=self,
+            date=datetime.utcnow(),
+            recipient_type=MessagingEvent.get_recipient_type_from_doc_type(recipient_doc_type),
+            recipient_id=recipient_id,
+            content_type=MessagingEvent.CONTENT_SMS,
+            case_id=case.get_id if case else None,
+            status=MessagingEvent.STATUS_IN_PROGRESS,
+        )
+        return obj
+
+    @property
+    def subevents(self):
```

Comment:
You could just use Django's related models features here instead of implementing it directly: https://docs.djangoproject.com/en/1.8/topics/db/queries/#backwards-related-objects